File: backend/services/alerts.js
const config = require("../../config/config.json");
const { getSystemStats } = require("./monitor");
const si = require("systeminformation");

let sendNotification = null;
try {
    const telegram = require("./telegram");
    sendNotification = telegram.sendNotification;
} catch (err) {
    // Telegram not configured
}

// Alert state tracking
const alertStates = {
    cpu: { triggered: false, lastAlert: 0 },
    memory: { triggered: false, lastAlert: 0 },
    disk: { triggered: false, lastAlert: 0 },
    temperature: { triggered: false, lastAlert: 0 },
    tunnel: { triggered: false, lastAlert: 0, lastRestart: 0, restartCount: 0 }
};

let monitoringInterval = null;

// Start alert monitoring
function startAlertMonitoring() {
    if (!config.alerts || !config.alerts.enabled) {
        console.log("⚠️  Alert monitoring disabled in config");
        return;
    }

    console.log("🔔 Starting alert threshold monitoring...");

    // Check every 60 seconds
    monitoringInterval = setInterval(async () => {
        await checkThresholds();
        await checkTunnelHealth();
    }, 60000);

    // Initial check
    checkThresholds();
    checkTunnelHealth();
}

// Stop alert monitoring
function stopAlertMonitoring() {
    if (monitoringInterval) {
        clearInterval(monitoringInterval);
        monitoringInterval = null;
        console.log("Alert monitoring stopped");
    }
}

// Check all thresholds
async function checkThresholds() {
    try {
        const stats = await getSystemStats();
        const thresholds = config.alerts.thresholds;
        const cooldown = config.alerts.cooldown || 300000; // 5 minutes default
        const now = Date.now();

        // CPU Check
        const cpuUsage = stats.cpu.usage;
        if (cpuUsage >= thresholds.cpu.critical) {
            await handleAlert("cpu", "critical", cpuUsage, thresholds.cpu.critical, "%", cooldown, now);
        } else if (cpuUsage >= thresholds.cpu.warning) {
            await handleAlert("cpu", "warning", cpuUsage, thresholds.cpu.warning, "%", cooldown, now);
        } else if (alertStates.cpu.triggered) {
            await handleRecovery("cpu", cpuUsage, "%");
        }

        // Memory Check
        const memUsage = stats.memory.usagePercent;
        if (memUsage >= thresholds.memory.critical) {
            await handleAlert("memory", "critical", memUsage, thresholds.memory.critical, "%", cooldown, now);
        } else if (memUsage >= thresholds.memory.warning) {
            await handleAlert("memory", "warning", memUsage, thresholds.memory.warning, "%", cooldown, now);
        } else if (alertStates.memory.triggered) {
            await handleRecovery("memory", memUsage, "%");
        }

        // Disk Check (primary disk)
        if (stats.disk && stats.disk.length > 0) {
            const primaryDisk = stats.disk[0];
            const diskUsage = primaryDisk.usagePercent;

            if (diskUsage >= thresholds.disk.critical) {
                await handleAlert("disk", "critical", diskUsage, thresholds.disk.critical, "%", cooldown, now);
            } else if (diskUsage >= thresholds.disk.warning) {
                await handleAlert("disk", "warning", diskUsage, thresholds.disk.warning, "%", cooldown, now);
            } else if (alertStates.disk.triggered) {
                await handleRecovery("disk", diskUsage, "%");
            }
        }

        // Temperature Check
        try {
            const temp = await si.cpuTemperature();
            if (temp.main && temp.main > 0) {
                const tempValue = Math.round(temp.main);

                if (tempValue >= thresholds.temperature.critical) {
                    await handleAlert("temperature", "critical", tempValue, thresholds.temperature.critical, "°C", cooldown, now);
                } else if (tempValue >= thresholds.temperature.warning) {
                    await handleAlert("temperature", "warning", tempValue, thresholds.temperature.warning, "°C", cooldown, now);
                } else if (alertStates.temperature.triggered) {
                    await handleRecovery("temperature", tempValue, "°C");
                }
            }
        } catch (err) {
            // Temperature not available, skip
        }

    } catch (error) {
        console.error("Alert monitoring error:", error.message);
    }
}

// Handle alert
async function handleAlert(metric, level, currentValue, threshold, unit, cooldown, now) {
    const state = alertStates[metric];

    // Check cooldown
    if (state.triggered && (now - state.lastAlert) < cooldown) {
        // Still in cooldown, skip
        return;
    }

    // Mark as triggered
    state.triggered = true;
    state.lastAlert = now;

    const emoji = level === "critical" ? "🔴" : "⚠️";
    const levelText = level.toUpperCase();

    const message = `${emoji} *${levelText}: High ${metric.charAt(0).toUpperCase() + metric.slice(1)} Usage*\n\n` +
        `${metric.toUpperCase()}: ${currentValue.toFixed(1)}${unit} (threshold: ${threshold}${unit})\n` +
        `Time: ${new Date().toLocaleTimeString()}\n\n` +
        `${level === "critical" ? "⚡ Action required!" : "Monitor closely"}`;

    console.log(`🔔 Alert: ${metric} ${level} - ${currentValue}${unit}`);

    if (sendNotification) {
        await sendNotification(message, level);
    }
}

// Handle recovery
async function handleRecovery(metric, currentValue, unit) {
    const state = alertStates[metric];
    state.triggered = false;

    const message = `✅ *Resolved: ${metric.charAt(0).toUpperCase() + metric.slice(1)} Back to Normal*\n\n` +
        `${metric.toUpperCase()}: ${currentValue.toFixed(1)}${unit}\n` +
        `Time: ${new Date().toLocaleTimeString()}`;

    console.log(`✅ Recovery: ${metric} - ${currentValue}${unit}`);

    if (sendNotification) {
        await sendNotification(message, "success");
    }
}

// Get alert status
function getAlertStatus() {
    return {
        enabled: config.alerts?.enabled || false,
        monitoring: monitoringInterval !== null,
        states: alertStates,
        thresholds: config.alerts?.thresholds || {}
    };
}

// Check Cloudflare tunnel health and auto-recover if needed
async function checkTunnelHealth() {
    const tunnelState = alertStates.tunnel;
    const now = Date.now();
    const cooldown = 120000; // 2 minutes between restarts
    const maxRestarts = 5; // Max restarts before giving up temporarily
    const resetAfter = 600000; // Reset restart count after 10 minutes of success

    try {
        // Try to get CF API status first
        const { getSetting } = require("./database");
        const cfConfig = getSetting('cloudflare');

        let isHealthy = false;
        let healthyCount = 0;
        let totalCount = 0;

        // Check via Cloudflare API if configured
        if (cfConfig && cfConfig.apiToken) {
            try {
                const cfService = require("./cloudflare");
                const tunnels = await cfService.listTunnels();

                if (tunnels && tunnels.length > 0) {
                    healthyCount = tunnels.filter(t => t.status === 'healthy').length;
                    totalCount = tunnels.length;
                    isHealthy = healthyCount === totalCount && totalCount > 0;
                }
            } catch (e) {
                // CF API error, skip this check
                return;
            }
        } else {
            // No CF API config, check local process
            const { getTunnelStatus } = require("./cloudflared");
            const status = await getTunnelStatus();
            isHealthy = status.processRunning;
        }

        // If healthy, reset state
        if (isHealthy) {
            if (tunnelState.triggered) {
                // Recovered!
                tunnelState.triggered = false;

                const message = `✅ *Tunnel Recovered*\n\nCloudflare tunnel is now healthy (${healthyCount}/${totalCount}).\n` +
                    `Time: ${new Date().toLocaleTimeString()}`;

                console.log(`✅ Tunnel recovered: ${healthyCount}/${totalCount} healthy`);

                if (sendNotification) {
                    await sendNotification(message, "success");
                }
            }

            // Reset restart count after period of success
            if (now - tunnelState.lastRestart > resetAfter) {
                tunnelState.restartCount = 0;
            }
            return;
        }

        // Tunnel is NOT healthy
        console.log(`⚠️ Tunnel unhealthy: ${healthyCount}/${totalCount}`);

        // Check if we're in cooldown
        if (now - tunnelState.lastRestart < cooldown) {
            return; // Wait for cooldown
        }

        // Check if we've hit max restarts
        if (tunnelState.restartCount >= maxRestarts) {
            if (!tunnelState.triggered) {
                tunnelState.triggered = true;
                tunnelState.lastAlert = now;

                const message = `🔴 *Tunnel Auto-Recovery Failed*\n\n` +
                    `Tunnel unhealthy after ${maxRestarts} restart attempts.\n` +
                    `Please check manually.\n` +
                    `Time: ${new Date().toLocaleTimeString()}`;

                console.log(`🔴 Tunnel auto-recovery failed after ${maxRestarts} attempts`);

                if (sendNotification) {
                    await sendNotification(message, "critical");
                }
            }
            return;
        }

        // Attempt to restart tunnel
        console.log(`🔄 Auto-restarting tunnel (attempt ${tunnelState.restartCount + 1}/${maxRestarts})...`);

        try {
            // Try systemd restart first on Linux
            if (process.platform === 'linux') {
                const { restartSystemdService } = require("./cloudflared");
                await restartSystemdService();
            } else {
                // Fallback: stop and start via cloudflared module
                const { stopTunnel, startTunnel } = require("./cloudflared");
                await stopTunnel();
                await new Promise(resolve => setTimeout(resolve, 2000));
                await startTunnel();
            }

            tunnelState.lastRestart = now;
            tunnelState.restartCount++;

            // Notify about restart attempt
            const message = `🔄 *Tunnel Auto-Restart*\n\n` +
                `Tunnel was unhealthy (${healthyCount}/${totalCount}), attempting restart...\n` +
                `Attempt: ${tunnelState.restartCount}/${maxRestarts}\n` +
                `Time: ${new Date().toLocaleTimeString()}`;

            console.log(`🔄 Tunnel restart initiated (attempt ${tunnelState.restartCount})`);

            if (sendNotification) {
                await sendNotification(message, "warning");
            }

        } catch (restartError) {
            console.error("Tunnel restart error:", restartError.message);
        }

    } catch (error) {
        console.error("Tunnel health check error:", error.message);
    }
}

module.exports = {
    startAlertMonitoring,
    stopAlertMonitoring,
    getAlertStatus,
    checkTunnelHealth
};

File: be/internal/alerts/alerts.go
// Package alerts ports backend/services/alerts.js as a background monitor.
package alerts

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"strings"
	"sync"
	"time"

	"github.com/nawfdev/home-panel/internal/config"
	"github.com/nawfdev/home-panel/internal/sysstats"
	"github.com/nawfdev/home-panel/internal/telegram"
	"github.com/nawfdev/home-panel/internal/tunnel"
)

type Config struct {
	Enabled    bool          `json:"enabled"`
	Cooldown   time.Duration `json:"-"`
	CooldownMS int64         `json:"cooldown"`
	Thresholds struct {
		CPU struct {
			Warning  float64 `json:"warning"`
			Critical float64 `json:"critical"`
		} `json:"cpu"`
		Memory struct {
			Warning  float64 `json:"warning"`
			Critical float64 `json:"critical"`
		} `json:"memory"`
		Disk struct {
			Warning  float64 `json:"warning"`
			Critical float64 `json:"critical"`
		} `json:"disk"`
		Temperature struct {
			Warning  float64 `json:"warning"`
			Critical float64 `json:"critical"`
		} `json:"temperature"`
	} `json:"thresholds"`
}

type state struct {
	Triggered    bool
	LastAlert    time.Time
	LastRestart  time.Time
	RestartCount int
}

type Service struct {
	cfg      Config
	telegram *telegram.Service
	tunnel   *tunnel.Service
	mu       sync.Mutex
	states   map[string]*state
}

func New(appCfg *config.Config, tg *telegram.Service, tun *tunnel.Service) *Service {
	cfg := defaultConfig()
	if len(appCfg.Alerts) > 0 {
		_ = json.Unmarshal(appCfg.Alerts, &cfg)
	}
	if cfg.CooldownMS <= 0 {
		cfg.CooldownMS = 300000
	}
	cfg.Cooldown = time.Duration(cfg.CooldownMS) * time.Millisecond
	return &Service{cfg: cfg, telegram: tg, tunnel: tun, states: map[string]*state{"cpu": {}, "memory": {}, "disk": {}, "temperature": {}, "tunnel": {}}}
}

func defaultConfig() Config {
	var cfg Config
	cfg.Enabled = false
	cfg.CooldownMS = 300000
	cfg.Thresholds.CPU.Warning = 70
	cfg.Thresholds.CPU.Critical = 85
	cfg.Thresholds.Memory.Warning = 75
	cfg.Thresholds.Memory.Critical = 90
	cfg.Thresholds.Disk.Warning = 80
	cfg.Thresholds.Disk.Critical = 90
	cfg.Thresholds.Temperature.Warning = 70
	cfg.Thresholds.Temperature.Critical = 85
	return cfg
}

func (s *Service) Start(ctx context.Context) {
	if !s.cfg.Enabled {
		log.Println("Alert monitoring disabled in config")
		return
	}
	if s.telegram != nil {
		s.telegram.SetMonitoring(true)
	}
	log.Println("Starting alert threshold monitoring...")
	go func() {
		s.check(ctx)
		ticker := time.NewTicker(time.Minute)
		defer ticker.Stop()
		for {
			select {
			case <-ctx.Done():
				if s.telegram != nil {
					s.telegram.SetMonitoring(false)
				}
				return
			case <-ticker.C:
				s.check(ctx)
			}
		}
	}()
}

func (s *Service) check(ctx context.Context) {
	stats, err := sysstats.GetSystemStats(ctx)
	if err != nil {
		log.Println("Alert monitoring error:", err)
		return
	}
	now := time.Now()
	s.threshold(ctx, "cpu", stats.CPU.Usage, s.cfg.Thresholds.CPU.Warning, s.cfg.Thresholds.CPU.Critical, "%", now)
	s.threshold(ctx, "memory", stats.Memory.UsagePercent, s.cfg.Thresholds.Memory.Warning, s.cfg.Thresholds.Memory.Critical, "%", now)
	if len(stats.Disk) > 0 {
		s.threshold(ctx, "disk", stats.Disk[0].UsagePercent, s.cfg.Thresholds.Disk.Warning, s.cfg.Thresholds.Disk.Critical, "%", now)
	}
	if temp := sysstats.GetTemperature(ctx); temp.Available && temp.Main != nil {
		s.threshold(ctx, "temperature", *temp.Main, s.cfg.Thresholds.Temperature.Warning, s.cfg.Thresholds.Temperature.Critical, "°C", now)
	}
	s.checkTunnel(ctx, now)
}

func (s *Service) threshold(ctx context.Context, metric string, current, warning, critical float64, unit string, now time.Time) {
	level := ""
	threshold := 0.0
	if critical > 0 && current >= critical {
		level, threshold = "critical", critical
	} else if warning > 0 && current >= warning {
		level, threshold = "warning", warning
	}
	if level == "" {
		s.recover(ctx, metric, current, unit)
		return
	}
	st := s.getState(metric)
	if st.Triggered && now.Sub(st.LastAlert) < s.cfg.Cooldown {
		return
	}
	st.Triggered = true
	st.LastAlert = now
	emoji := "⚠️"
	if level == "critical" {
		emoji = "🔴"
	}
	msg := fmt.Sprintf("%s *%s: High %s Usage*\n\n%s: %.1f%s (threshold: %.0f%s)\nTime: %s", emoji, upper(level), title(metric), upper(metric), current, unit, threshold, unit, now.Format("15:04:05"))
	log.Printf("Alert: %s %s - %.1f%s", metric, level, current, unit)
	s.notify(ctx, msg)
}

func (s *Service) recover(ctx context.Context, metric string, current float64, unit string) {
	st := s.getState(metric)
	if !st.Triggered {
		return
	}
	st.Triggered = false
	msg := fmt.Sprintf("✅ *Resolved: %s Back to Normal*\n\n%s: %.1f%s\nTime: %s", title(metric), upper(metric), current, unit, time.Now().Format("15:04:05"))
	log.Printf("Recovery: %s - %.1f%s", metric, current, unit)
	s.notify(ctx, msg)
}

func (s *Service) checkTunnel(ctx context.Context, now time.Time) {
	st := s.getState("tunnel")
	status := s.tunnel.GetStatus(ctx)
	if status.ProcessRunning || status.IsReady {
		if st.Triggered {
			st.Triggered = false
			s.notify(ctx, "✅ *Tunnel Recovered*\n\nCloudflare tunnel is now healthy.\nTime: "+now.Format("15:04:05"))
		}
		if now.Sub(st.LastRestart) > 10*time.Minute {
			st.RestartCount = 0
		}
		return
	}
	if now.Sub(st.LastRestart) < 2*time.Minute {
		return
	}
	if st.RestartCount >= 5 {
		if !st.Triggered {
			st.Triggered = true
			st.LastAlert = now
			s.notify(ctx, "🔴 *Tunnel Auto-Recovery Failed*\n\nTunnel unhealthy after 5 restart attempts.\nPlease check manually.\nTime: "+now.Format("15:04:05"))
		}
		return
	}
	_, _ = s.tunnel.StartTunnel(ctx)
	st.LastRestart = now
	st.RestartCount++
	s.notify(ctx, fmt.Sprintf("🔄 *Tunnel Auto-Restart*\n\nTunnel was unhealthy, attempting restart...\nAttempt: %d/5\nTime: %s", st.RestartCount, now.Format("15:04:05")))
}

func (s *Service) getState(metric string) *state {
	s.mu.Lock()
	defer s.mu.Unlock()
	st := s.states[metric]
	if st == nil {
		st = &state{}
		s.states[metric] = st
	}
	return st
}

func (s *Service) notify(ctx context.Context, msg string) {
	if s.telegram == nil {
		return
	}
	if err := s.telegram.SendNotification(ctx, msg); err != nil {
		log.Println("Telegram notification failed:", err)
	}
}

func title(s string) string {
	if s == "" {
		return s
	}
	return strings.ToUpper(s[:1]) + s[1:]
}

func upper(s string) string { return strings.ToUpper(s) }
